Sources/LiveKit/Support/FFTProcessor.swift

```diff
@@ -0,0 +1,188 @@
+                bandMagnitudes[i] = magnitudes[magsStartIdx]
+            }
+
+            // Compute average frequency
+            bandFrequencies[i] = _averageFrequencyInRange(magsStartIdx, magsEndIdx, sampleRate: sampleRate)
+        }
+
+        return FFTComputeBandsResult(count: bandsCount, magnitudes: bandMagnitudes, frequencies: bandFrequencies)
+    }
+
+    @inline(__always) private func _magnitudeIndex(for frequency: Float, sampleRate: Float) -> Int {
+        Int(Float(magnitudes.count) * frequency / sampleRate.nyquistFrequency)
+    }
+
+    @inline(__always) private func _computeAverage(_ array: [Float], _ startIdx: Int, _ stopIdx: Int) -> Float {
+        var mean: Float = 0
+        let count = stopIdx - startIdx
+        array.withUnsafeBufferPointer { bufferPtr in
+            let ptr = bufferPtr.baseAddress! + startIdx
+            vDSP_meanv(ptr, 1, &mean, UInt(count))
+        }
+        return mean
+    }
+
+    @inline(__always) private func _computeBandwidth(for sampleRate: Float) -> Float {
+        sampleRate.nyquistFrequency / Float(magnitudes.count)
+    }
+
+    @inline(__always) private func _averageFrequencyInRange(_ startIndex: Int, _ endIndex: Int, sampleRate: Float) -> Float {
+        let bandwidth = _computeBandwidth(for: sampleRate)
+        return (bandwidth * Float(startIndex) + bandwidth * Float(endIndex)) / 2
+    }
+}
+
+class FFTProcessor {
+    public enum WindowType {
+        case none
+        case hanning
+        case hamming
+    }
+
+    public enum ScaleType {
+        case linear
+        case logarithmic
+    }
+
+    public let bufferSize: Int
+    public let windowType: WindowType
+    public let scaleType: ScaleType
+
+    private let bufferHalfSize: Int
+    private let bufferLog2Size: Int
+    private var window: [Float] = []
+    private var fftSetup: FFTSetup
+    private var complexBuffer: DSPSplitComplex
+    private var realPointer: UnsafeMutablePointer<Float>
+    private var imaginaryPointer: UnsafeMutablePointer<Float>
+    private var zeroDBReference: Float = 1.0
+
+    init(bufferSize: Int, scaleType: ScaleType = .linear, windowType: WindowType = .hanning) {
+        self.bufferSize = bufferSize
+        self.scaleType = scaleType
+        self.windowType = windowType
+
+        bufferHalfSize = bufferSize / 2
+        bufferLog2Size = Int(log2f(Float(bufferSize)))
+
+        fftSetup = vDSP_create_fftsetup(UInt(bufferLog2Size), FFTRadix(FFT_RADIX2))!
+
+        realPointer = .allocate(capacity: bufferHalfSize)
+        imaginaryPointer = .allocate(capacity: bufferHalfSize)
+
+        realPointer.initialize(repeating: 0.0, count: bufferHalfSize)
+        imaginaryPointer.initialize(repeating: 0.0, count: bufferHalfSize)
+
+        complexBuffer = DSPSplitComplex(realp: realPointer, imagp: imaginaryPointer)
+        setupWindow()
+    }
+
+    deinit {
+        vDSP_destroy_fftsetup(fftSetup)
+        realPointer.deallocate()
+        imaginaryPointer.deallocate()
```

Comment:
is there a reason you're managing raw pointers in this class rather than just using arrays and nested withUnsafeMutableBufferPointer calls as in the apple sample code? looks scary to me :shrug: https://developer.apple.com/documentation/accelerate/vdsp/fast_fourier_transforms/finding_the_component_frequencies_in_a_composite_sine_wave